### ISSUE TITLE
Stuck while open large repo

### ISSUE THREAD

**Issue by deitry** [eamodio/vscode-gitlens#581]
- GitLens Version: 9.0.3
- VSCode Version: 1.29.1, 1.30 Insiders
- OS Version: Kubuntu 18.04.1

After I did update GitLens from 8.x to 9.x version, I decided to test it on project which contains about 8000 commits and huge amount of files. 

GitLens debug log:
`[2018-12-07 08:08:49:735] [1] GitService.initialize
[2018-12-07 08:08:49:892] Git found: 2.17.1 @ /usr/bin/git • 101 ms
[2018-12-07 08:08:49:892] [1] GitService.initialize completed • 156 ms
[2018-12-07 08:08:49:898] Starting repository search in 1 folders
[2018-12-07 08:08:49:899] Searching for repositories (depth=1) in '/home/user/workspace/toolchain/pkg/project' ...
[2018-12-07 08:08:49:918] [2] GitService.getRepoPath — e=Uri(file:///home/user/.config/Code/User/settings.json)
[2018-12-07 08:08:49:919] [3] GitService.getRepository — e=Uri(file:///home/user/.config/Code/User/settings.json), t={"skipCacheUpdate":true}
[2018-12-07 08:08:49:941] GitLens(v9.0.3) activated • 223 ms
[2018-12-07 08:08:50:129] [/home/user/workspace/toolchain/pkg/project] git rev-parse --show-toplevel • 229 ms 
[2018-12-07 08:08:50:134] Repository found in '/home/user/workspace/toolchain/pkg/project'
[2018-12-07 08:08:55:724] [4] GitService.getOrderedRepositories
[2018-12-07 08:08:55:724] [5] GitService.getRepositories
[2018-12-07 08:08:55:975] [6] RepositoriesNode.onVisibilityChanged — e={"visible":true}
[2018-12-07 08:08:55:975] [7] RepositoriesNode.ensureSubscription
[2018-12-07 08:08:55:976] [8] RepositoriesNode.unsubscribe
[2018-12-07 08:08:55:976] [9] RepositoriesNode.triggerChange
[2018-12-07 08:08:55:976] [a] RepositoriesView.refreshNode — e=RepositoriesNode
[2018-12-07 08:08:55:976] [b] RepositoriesNode.refresh
[2018-12-07 08:08:55:976] [6] RepositoriesNode.onVisibilityChanged completed • 0 ms
[2018-12-07 08:08:55:976] [8] RepositoriesNode.unsubscribe completed • 0 ms
[2018-12-07 08:08:55:976] [b] RepositoriesNode.refresh completed • 0 ms
[2018-12-07 08:08:55:977] [7] RepositoriesNode.ensureSubscription completed • 1 ms
[2018-12-07 08:08:55:977] [c] RepositoriesView.triggerNodeChange — e=RepositoriesNode
[2018-12-07 08:08:55:977] [c] RepositoriesView.triggerNodeChange completed • 0 ms
[2018-12-07 08:08:55:977] [a] RepositoriesView.refreshNode completed • 0 ms
[2018-12-07 08:08:55:977] [9] RepositoriesNode.triggerChange completed • 0 ms
[2018-12-07 08:09:47:454] [d] RepositoriesNode.onVisibilityChanged — e={"visible":false}
[2018-12-07 08:09:47:454] [e] RepositoriesNode.ensureSubscription
[2018-12-07 08:09:47:455] [f] RepositoriesNode.unsubscribe
[2018-12-07 08:09:47:455] [d] RepositoriesNode.onVisibilityChanged completed • 0 ms
[2018-12-07 08:09:47:455] [f] RepositoriesNode.unsubscribe completed • 0 ms
[2018-12-07 08:09:47:455] [e] RepositoriesNode.ensureSubscription completed • 0 ms
[2018-12-07 08:09:47:916] [10] RepositoriesNode.onVisibilityChanged — e={"visible":true}
[2018-12-07 08:09:47:917] [11] RepositoriesNode.ensureSubscription
[2018-12-07 08:09:47:917] [12] RepositoriesNode.unsubscribe
[2018-12-07 08:09:47:917] [13] RepositoriesNode.triggerChange
[2018-12-07 08:09:47:917] [14] RepositoriesView.refreshNode — e=RepositoriesNode
[2018-12-07 08:09:47:919] [15] RepositoriesNode.refresh
[2018-12-07 08:09:47:919] [10] RepositoriesNode.onVisibilityChanged completed • 2 ms
[2018-12-07 08:09:47:919] [12] RepositoriesNode.unsubscribe completed • 1 ms
[2018-12-07 08:09:47:919] [15] RepositoriesNode.refresh completed • 0 ms
[2018-12-07 08:09:47:919] [11] RepositoriesNode.ensureSubscription completed • 2 ms
[2018-12-07 08:09:47:919] [16] RepositoriesView.triggerNodeChange — e=RepositoriesNode
[2018-12-07 08:09:47:920] [16] RepositoriesView.triggerNodeChange completed • 0 ms
[2018-12-07 08:09:47:920] [14] RepositoriesView.refreshNode completed • 1 ms
[2018-12-07 08:09:47:920] [13] RepositoriesNode.triggerChange completed • 2 ms`

After that, progress bar in Repositories side bar is just running forever. Therefore, none of GitLens functions are available. Maybe it can somehow complete refreshing, but after 15 minutes I consider it a regression in compare to previous GitLens version. In previous version it refreshed in _seconds_. 

I tested GitLens on other repositories (like my own toolchain or https://github.com/Guake/guake) and it seems to work fine.

**Comment by eamodio** [comment]
@deitry Is this repo accessible? I've tried with some large repos ([git](https://github.com/git/git), [vscode](https://github.com/Microsoft/vscode), etc) and I haven't seen any performance issue like you are describing. I would love to get to the bottom of what you are seeing.

Can you please enable output channel logging by setting `"gitlens.outputLevel": "debug"` in your `settings.json`. This will enable logging to the *GitLens* and *GitLens (Git)* channels in the *Output* pane. Once enabled, please reproduce the issue (if possible) and attach the log lines from the *GitLens* and *GitLens (Git)* channels.

**Comment by deitry** [comment]
@eamodio No, it's not accessible. We have internal server with some repos like project and toolchain. We have following structure: `~/workspace/toolchain [repo]/pkg/project [repo]/external_libs [repo]/[submodule for each lib]`
As I said, GitLens with toolchain repo works fine, as well as external libs repo with submodules, but for project it fails.

Log from GitLens Output with `"gitlens.outputLevel": "debug"` is shown above, log from `GitLens (Git)` is below (yes, it's only this single line):
> [2018-12-07 09:06:18:065] [/home/user/workspace/toolchain/pkg/project] git rev-parse --show-toplevel • 106 ms

**Comment by deitry** [comment]
That is sooo strange. It reproduces always and only on this single repo. Any other opens as usual. But most important, it works perfectly even on another clone of this repo! I tried to remove nested external_libs repository, but it doens't solve issue.

Could it be something inside working folder?

**Comment by deitry** [comment]
Ooh, I get it!
For toolchain purposes there was a symlink that pointed to the project itself (I don't know really why it's needed, I didn't do that): `../project/src -> ../project`

As soon as I delete symlink, GitLens started to operate normally. I guess GitLens 8 doesn't followed symlinks since it worked well with that.

So is it correct behavior?

**Comment by eamodio** [comment]
@deitry so glad you figured it out! I don't know that I changed anything on purpose to follow symlinks -- as I'm not really sure where things went awry. But I think following the symlinks is probably ultimately desirable.
